File: mwc/MWC_15_C7_P3_BLUE_AND_GREEN.cpp
#include <stdio.h>
#include <string.h>

#define scan(x) do{while((x=getchar())<'0'); for(x-='0'; '0'<=(_=getchar()); x=(x<<3)+(x<<1)+_-'0');}while(0)
char _;

int cache [1048576]; //1 << 20
unsigned int N, S, L, R, I, N_MINUS_ONE, num, MASK;

inline void ways (unsigned int marbles, int M) {
	if (cache [marbles] < M) {
		if (cache [marbles] == -1) {
			++num;
		}
		cache [marbles] = M;
		
		if (M >= S) {
			for (int n = 0; n < N; ++n) {
				ways (marbles ^ (1 << n), M - S);
			}
		}
		
		if (M >= R) {
			ways (((marbles >> 1) | (marbles << N_MINUS_ONE)) & MASK, M - R);
		}
		
		if (M >= L) {
			ways (((marbles << 1) | (marbles >> N_MINUS_ONE)) & MASK, M - L);
		}
		
		if (M >= I) {
			ways (~marbles & MASK, M - I);
		}
	}
}

int main () {
	int M;
	scan (N); scan (M);
	
	MASK = (1 << N) - 1;
	N_MINUS_ONE = N - 1;//what is this optimization
	
	scan (S); scan (L); scan (R); scan (I);
	
	memset (cache, -1, sizeof (cache));
	
	unsigned int start = 0;
	
	for (int n = 0; n < N; ++n) {
		if (getchar () == 'G') {
			start ^= (1 << (N_MINUS_ONE - n));
		}
	}
	
	ways (start, M);
	
	printf ("%d", num);
}

File: mwc/MWC_15_C7_P3_BLUE_AND_GREEN.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

// atharva washimkar
// August 11, 2018

public class MWC_15_C7_P3_BLUE_AND_GREEN {

	static int[] cache;//stores max money of each position visited -1 = unvisited, anything higher means visited
	static int N, S, L, R, I, MASK, N_MINUS_ONE;
	static int num = 0;//unique marbles

	public static void ways (int marbles, int M) {
		if (cache[marbles] < M) {//new config OR old config but new options
			//System.out.println (Integer.toBinaryString (marbles));

			if (cache[marbles] == -1) {//only increment if new config
				num++;
			}
			cache[marbles] = M;//avoid visiting unless you have more than M money

			if (M >= R) {
				ways (((marbles >>> 1) | ((marbles & (MASK >> N_MINUS_ONE)) << N_MINUS_ONE)) & MASK, M - R);//rotate
				// right
			}

			if (M >= L) {
				ways (((marbles << 1) | (marbles >> N_MINUS_ONE)) & MASK, M - L);//rotate left
			}

			if (M >= I) {
				ways (~marbles & MASK, M - I);//invert
			}

			if (M >= S) {
				for (int n = 0; n < N; n++) {
					ways (marbles ^ (1 << n), M - S);//set bit
				}
			}
		}
	}

	public static void main (String[] t) throws IOException {
		BufferedReader in = new BufferedReader (new InputStreamReader (System.in));
		t = in.readLine ().split (" ");

		N = Integer.parseInt (t[0]);
		int M = Integer.parseInt (t[1]);
		MASK = Integer.MAX_VALUE >> (31 - N);
		N_MINUS_ONE = N - 1;
		cache = new int[1 << N];

		t = in.readLine ().split (" ");
		S = Integer.parseInt (t[0]);
		L = Integer.parseInt (t[1]);
		R = Integer.parseInt (t[2]);
		I = Integer.parseInt (t[3]);

		Arrays.fill (cache, -1);

		String m = in.readLine ();
		int start = 0;

		for (int n = 0; n < N; n++) {
			if (m.charAt (n) == 'B') {
				start ^= (1 << (N_MINUS_ONE - n));
			}
		}

		ways (start, M);
		System.out.println (num);
	}
}